### PR TITLE
Cuda dev

### PR DESCRIPTION
**Pull request by antmedellin**

### CODE REVIEW

File: hyperfunctions.cpp [hunk 1]
```diff
@@ -755,14 +836,17 @@ void  HyperFunctions::SemanticSegmenter()
             double low_val;
             for (int i=0; i<temp_results.size(); i++)
             {
+
                 if (i==0)
                 {
-                low_val=temp_results[i].at<uchar>(j,k);
-                temp_class_img.at<Vec3b>(Point(k,j)) = reference_colors[i];
+                    low_val=temp_results[i].at<uchar>(j,k);
```
Comment (antmedellin) [inline hyperfunctions.cpp:842]: why is index not the same low_val and temp_class_img??

File: hyperfunctions.cpp [hunk 2]
```diff
@@ -207,13 +207,15 @@ void  HyperFunctions::DispClassifiedImage()
...
    Mat temp_img;
+
    cv::resize(false_img,temp_img,Size(WINDOW_WIDTH, WINDOW_HEIGHT),INTER_LINEAR); 
    imshow("False Image", temp_img);
 }
 
 void  HyperFunctions::DispSpecSim()
 {
    Mat temp_img;
+   spec_simil_img.convertTo(spec_simil_img, CV_8UC1);
```
Comment (antmedellin) [inline hyperfunctions.cpp:218]: conversion should be done in different location

File: hypergpufunctions.h [hunk 3]
```diff
@@ -34,18 +34,20 @@ class HyperFunctionsGPU : public HyperFunctions {
...
     int grid_size ;
+    int *img_array_base;
     
     
     void spec_sim_GPU();
     void allocate_memory(int* img_array);
     void deallocate_memory( );
     int* mat_to_oneD_array_parallel_parent();
-    void oneD_array_to_mat(int* img_array);
+    void oneD_array_to_mat(int* img_array, int rows, int cols);
```
Comment (antmedellin) [inline hypergpufunctions.h:44]: is it necessary to pass rows, cols? Can it be figured out from inside function?

File: hypergpufunctions.cu [hunk 4]
```diff
@@ -140,58 +189,77 @@ void HyperFunctionsGPU::spec_sim_GPU() {
...
     num_lay=  mlt1.size();
     block_size = 512;
-    grid_size = ((N_points + block_size) / block_size);
-
+    grid_size = ((N_points + block_size) / block_size); 
+    
     int tmp_len1=reference_spectrums[ref_spec_index].size(); 
 
-    cudaHostAlloc ((void**)&out, sizeof(int) *N_size, cudaHostAllocDefault);
+    cudaHostAlloc ((void**)&out, sizeof(unsigned int) * N_size, cudaHostAllocDefault);
```
Comment (antmedellin) [inline hypergpufunctions.cu:218]: what affect does changing to unsigned int have?

File: hypergpufunctions.cu [hunk 5]
```diff
@@ -140,58 +189,77 @@ void HyperFunctionsGPU::spec_sim_GPU() {
...
         for (int j=0; j<spec_simil_img.rows; j++)
         {
-            spec_simil_img.at<uchar>(j,k)=img_array[val_it];
+            spec_simil_img.at<uchar>(j,k)=img_array[val_it]; 
             val_it+=1;
+            //Conversion of 1-d array containing pixel values to 8 bit one channel 2-d OPENCV matrix
         }
-    }
+    }*/
+    spec_simil_img = cv::Mat(rows, cols, CV_32SC1, img_array);
```
Comment (antmedellin) [inline hypergpufunctions.cu:254]: can we use mlt1[0].rows / cols to figure out rows and cols, so we do not need to pass the values?

Comment (antmedellin) [reply hypergpufunctions.cu:254]: spec_simil_img.convertTo(spec_simil_img, CV_8UC1);
conversion should also be done here

File: hypergpufunctions.cu [hunk 6]
```diff
@@ -201,36 +269,32 @@ void HyperFunctionsGPU::oneD_array_to_mat(int* img_array, int cols, int rows, in
...
-
-                Vec3b color = classified_img.at<Vec3b>(Point(0,0));
-                color[0]=img_array[val_it] ;
-                color[1]= img_array[val_it+1] ;
-                color[2]=img_array[val_it+2] ;
-                //cout<<color<<endl;
-                classified_img.at<Vec3b>(Point(k,j))= color   ;
-                val_it+=3;
-
+            Vec3b color = classified_img.at<Vec3b>(Point(0,0));
```
Comment (antmedellin) [inline hypergpufunctions.cu:272]: can we do the same casting here as we did for greyscale images?

File: hypergpufunctions.cu [hunk 7]
```diff
@@ -267,93 +331,134 @@ int* HyperFunctionsGPU::mat_to_oneD_array_parallel_parent(vector<Mat>* matvector
...
-    cudaMalloc((void**)&d_clasified_img_array, sizeof(int) * N_points_sim);
+
+    /**
+     * out2 : 1-d array that represents matrix with 3 channels, to store R G and B values. 
+     * d_clasified_img_array : 1-d array containing spectra similarity values 
+     * d_color_info : holds reference colors that will be used to colorize our final image
+     * 
+    */
+   
+    cudaHostAlloc ((void**)&out2, sizeof(int) *N_size_sim, cudaHostAllocDefault);
```
Comment (antmedellin) [inline hypergpufunctions.cu:434]: int vs unsigned int here?

File: hypergpufunctions.cu [hunk 8]
```diff
@@ -267,93 +331,134 @@ int* HyperFunctionsGPU::mat_to_oneD_array_parallel_parent(vector<Mat>* matvector
...
+        //updating the memory allocated on the GPU that stores the current reference spectra
+        cudaMemcpy(d_ref_spectrum, ref_spectrum, sizeof(int) * tmp_len1, cudaMemcpyHostToDevice); 
         this->spec_sim_GPU();
         similarity_images.push_back(spec_simil_img);
     }
 
-    this->deallocate_memory();
+    this->deallocate_memory(); //deallocate memory on GPU used for spectral similarity algorithms
     int array_size2=similarity_images[1].rows*similarity_images[1].cols*similarity_images.size();  
+    int mat_size = similarity_images[1].rows * similarity_images[1].cols;
```
Comment (antmedellin) [inline hypergpufunctions.cu:411]: are mat_size and mat_cols used?

File: hypergpufunctions.h (hunk 3 above)
Comment (Leofeng7) [reply hypergpufunctions.h:44]: fixed

File: hypergpufunctions.cu (hunk 4 above)
Comment (Leofeng7) [reply hypergpufunctions.cu:218]: woops I was trying to make casting to 8UC1 work, forgot to change this back

File: hypergpufunctions.cu (hunk 5 above)
Comment (Leofeng7) [reply hypergpufunctions.cu:254]: fixed

File: hypergpufunctions.cu (hunk 8 above)
Comment (Leofeng7) [reply hypergpufunctions.cu:411]: removed

File: hyperfunctions.cpp (hunk 2 above)
Comment (Leofeng7) [reply hyperfunctions.cpp:218]: fixed

File: hyperfunctions.cpp (hunk 1 above)
Comment (Leofeng7) [reply hyperfunctions.cpp:842]: Wait anthony I don't think I changed anything in this function

**Review by Leofeng7 — COMMENTED** [review]
Looks good

File: hypergpufunctions.cu (hunk 6 above)
Comment (Leofeng7) [reply hypergpufunctions.cu:272]: added

File: hyperfunctions.cpp (hunk 1 above)
Comment (antmedellin) [reply hyperfunctions.cpp:841]: yeah this was me. It was a reminder to myself to look into

**Pull request merged**